hadoop-hdds/docs/content/design/s3-performance.md

```diff
@@ -0,0 +1,205 @@
+
+## Caching
+
+Hadoop RPC has a very specific caching layer which is **not used** by Ozone S3g. This section describe the caching of the Hadoop RPC, but safe to skip (It explain how is the caching ignored).
+
+As creating new Hadoop RPC connection is an expensive operation Hadoop RPC has an internal caching mechanism to cache client and connections (!). This caching is hard-coded and based on static fields (couldn't be adjusted easily).
+
+Hadoop RPC client is usually created by `RPC.getProcolProxy`. For example:
+
+```
+HelloWorldServicePB proxy = RPC.getProtocolProxy(
+            HelloWorldServicePB.class,
+            scmVersion,
+            new InetSocketAddress(1234),
+            UserGroupInformation.getCurrentUser(),
+            configuration,
+            new StandardSocketFactory(),
+            Client.getRpcTimeout(configuration),
+            retryPolicy).getProxy();
+```
+
+This code fragment creates a new client which can be used from the code, and it uses multiple caches for client creation.
+
+1. Protocol engines are cached by `RPC.PROTOCOL_ENGINES` static field, but it's safe to assume that the `ProtobufRpcEngine` is used for most of the current applications.
+
+2. `ProtobufRpcEngine` has a static `ClientCache` field which caches the client instances with the `socketFactory` and `protocol` as the key.
+
+3. Finally the `Client.getConnection` method uses a cache to cache the connections:
+
+   ```
+   connection = connections.computeIfAbsent(remoteId,
+       id -> new Connection(id, serviceClass, removeMethod));
+   ```
+
+   The key for the cache is the `remoteId` which includes all the configuration, connection parameters (like destination host) and `UserGroupInformation` (UGI).
+
+The caching of the connections can cause very interesting cases. As an example, let's assume that delegation token is invalidated with an RPC call. The workflow can be something like this:
+
+1. create protocol proxy (with token authentication)
+2. invalidate token (rpc call)
+3. close protocol proxy (connection may not be closed. depends from the cache)
+4. create a new protocol proxy
+5. If connection is cached (same UGI) services can be used even if the token is invalidated earlier (as the token is checked during the initialization of the tokens).
+
+Fortunately this behavior doesn't cause any problem in case of Ozone and S3g. UGI (which is part of the cache key of the connection cache) equals if (and only if) the underlying `Subject` is the same.
```

Comment:
Just to clarify my understanding, we don't have invalidate token method for s3G right?

As this token is generated from Client auth header fields. Means token is generated per request.

We are sending the required info to validate the auth header with the secret which OM has.

Reply:
> Just to clarify my understanding, we don't have invalidate token method for s3G right?

You are right. This example is independent of s3g just explains how the cache works. I tried to describe the problem with the simple delegation token. (I can add it as a note)